### PR TITLE
fix(crd): make the spec object required across all the CRDs (except KafkaProxy).  KafkaProxy without specs are deprecated.

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Bugfix

### Description

I think owing to an oversight, all our API are accepting APIs object with `spec` field absent.  In all cases **except** `KafkaProxy`, this is meaningless, and leads to a reconciliation failure. We saw a case on #3330 (although it is not clear why the issue chose that moment to appear).

~~KafkaProxy without `spec` is accepted, and is treated the same as an empty spec, which is legal.  Whilst I am not sure it was the original intention was to accept a `KafkaProxy` without spec, continuing to accept it now seems the best, most compatible, path.~~

The alternative approach would be to have the KafkaProxyReconciler warn about the missing spec field (IOW deprecate its absence) in the object to encourage users to add it, and make the field mandatory after the deprecation period is over.  

EDIT: After discussion on this PR, we decided that deprecating KafkaProxies without `spec` fields is the way to go.

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] PR raised from a fork of this repository and made from a branch rather than main. 
- [x] Write tests
- [ ] Update documentation
- [x] Make sure all unit/integration tests pass
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, make sure system tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [x] Ensure the PR references relevant issue(s) so they are closed on merging.
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

### CODE REVIEW

**Comment by tombentley** [comment]
> The alternative approach would be to have the KafkaProxyReconciler warn about the missing spec field (IOW deprecate its absence) in the object to encourage users to add it, and make the field mandatory after the deprecation period is over.

I'd be in favour of this, just so we don't have to carry around this needless inconsistency for ever.

File: kroxylicious-kubernetes-api/src/main/resources/META-INF/fabric8/kafkaproxies.kroxylicious.io-v1.yml [hunk 1]
```diff
@@ -228,3 +228,4 @@ spec:
                   description: "The number of instances of the proxy pods that have reported ready."
                   type: integer
                   format: int32
+          required: [] # Absent spec is treated as if it were present and empty
```
Comment (SamBarker) [inline kroxylicious-kubernetes-api/src/main/resources/META-INF/fabric8/kafkaproxies.kroxylicious.io-v1.yml]:
```suggestion
          required: [] #(deprecated) Absent spec is treated as if it were present and empty
```

Lets mark it as deprecated so we can mark it required in the v1 api

**Review by SamBarker — APPROVED** [review]
Relatedly should we think about reving the API version?

**Comment by tombentley** [comment]
> Relatedly should we think about reving the API version?

My experience from Strimzi (`v1beta1` to `v1beta2`, IIRC) is that this is pretty painful, both for the developers (us) and for users. Since Strimzi are moving `to` v1 APIs at the moment it might be worth syncing with them to find out how they're doing it this time (maybe they have a better way now?). But I certainly wouldn't be in a rush to bump the API version until you're really sure that there's nothing else you want to get rid of. My feeling is that people are starting to use the operator, and we should wait for at least a year or two to let the problems come out before we contemplate bumping the api version.

**Comment by SamBarker** [comment]
> > Relatedly should we think about reving the API version?
> 
> My experience from Strimzi (`v1beta1` to `v1beta2`, IIRC) is that this is pretty painful, both for the developers (us) and for users. Since Strimzi are moving `to` v1 APIs at the moment it might be worth syncing with them to find out how they're doing it this time (maybe they have a better way now?). But I certainly wouldn't be in a rush to bump the API version until you're really sure that there's nothing else you want to get rid of. My feeling is that people are starting to use the operator, and we should wait for at least a year or two to let the problems come out before we contemplate bumping the api version.

Ohh right! I'd kind of assumed we'd go v1 on this API with the runtime v1.0.0 but we deliberately decoupled them.

**Comment by k-wall** [comment]
Thanks for the feedback.  I'll add a warning to KafkaProxyReconciler to warn if the reconciler encounters a resource without a spec.  I'll also update tests and test-data to pass an empty spec.  I will of course keep a test that verifies that a KafkaProxy without spec is tolerated.

File: kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconciler.java [hunk 2]
```diff
@@ -144,6 +151,11 @@ public class KafkaProxyReconciler implements
     private final SecureConfigInterpolator secureConfigInterpolator;
     private final KafkaProxyStatusFactory statusFactory;
 
+    private final Cache<String, Boolean> resourcesWithAbsentSpecs = Caffeine.newBuilder()
```
Comment (k-wall) [inline kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconciler.java:154]: I want to avoid the (remote) possibility of the cache turning into a memory leak.  I initially looked at [JOSDK Cleaner API](https://javaoperatorsdk.io/docs/documentation/reconciler/#:~:text=trigger%20reconciliation%20requests.-,Implementing%20Reconciler%20and%20Cleaner%20Interfaces,-To%20implement%20a) utilising the `#cleanup` method to clean-up the cache when a resource is deleted.  That seemed like a like a significant change/risk without really bringing to much benefit (after all - this is a deprecation message).  In the end I decided on Caffeine cache with a 1 hour expiry.  This'll nag the user at most once per hour and avoids the possibility of a leak in all but extreme use-cases.

File: kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconciler.java [hunk 3]
```diff
@@ -465,6 +478,22 @@ public List<EventSource<?, KafkaProxy>> prepareEventSources(EventSourceContext<K
                 buildKafkaProxyIngressEventSource(context));
     }
 
+    @VisibleForTesting
+    void reportAbsentSpecIfNecessary(KafkaProxy proxy, Logger log) {
```
Comment (k-wall) [inline kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconciler.java:482]: I started off using logcaptor, but got in a classpath minefield between the logging dependencies we want for production and the logging dependencies need for logcaptor.  I fell back on mocking the Logger and making the method visible for test.

**Comment by k-wall** [comment]
@SamBarker could I ask for a re-review on this one please?

File: kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconciler.java [hunk 4]
```diff
@@ -465,6 +478,22 @@ public List<EventSource<?, KafkaProxy>> prepareEventSources(EventSourceContext<K
...
+        var resourceUid = Optional.of(proxy).map(HasMetadata::getMetadata).map(ObjectMeta::getUid);
+        resourceUid.ifPresent(uid -> {
+            if (proxy.getSpec() == null) {
+                if (resourcesWithAbsentSpecs.asMap().putIfAbsent(uid, true) == null) {
+                    log.warn("{} has no spec, please add an empty one. "
+                            + " Support for spec-less KafkaProxy resources is deprecated and will be removed in a future release.", ResourcesUtil.namespacedSlug(proxy));
```
Comment (tombentley) [inline kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconciler.java:488]: The problem with using logging as the reporting mechanism is we don't know whether the people with access to the logs are the same people who manage the CR. Often they are, but in general they can be distinct sets of people.

I think it's very likely that we will end up wanting to poke the people who own the CRs to make changes to them again the future, so it would be good to establish a good mechanism for doing that now, and using it consistently. (This is the path we trod in Strimzi).

Putting these two points together, I think we need should be able to report non-fatal problems with a CR on the CR itself, in the status section (in addition to the logging, which I think is a useful extra line of communication). For example we could add something like `type: DeprecationWarning, status: True, message: Future versions of this API will require 'spec' to be present` to the `status.conditions`.

Assuming we were to do this, I would decouple the logging in the following way:

* Attach these conditions to the `KafkaProxyContext` at the point they're detected.
* Copy them to the status when computing the status patch, and log at this point too. That would allow you do log all the deprecation warnings at once using a single cache. But you could probably avoid the cache entirely by using the timestamp on the condition to determine when you last logged.

Comment (k-wall) [reply kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconciler.java:488]: Good point.  I agree using a condition is better. However, would you mind if I address this separately?  

I'd like to get 0.19.0 out soon and I wish to include this defect fix. 

I'd like to merge this PR as is and raise an issue to report the deprecation issue via the status.

**Review by tombentley — APPROVED** [review]
Thanks Keith
